### DISCUSSION TITLE
"Writing characteristic failed with status code 200."

### DISCUSSION THREAD

**Discussion by BAyotte** [capacitor-community/bluetooth-le#680] [Q&A]
Hi, I have been trying to communicate with a BLE device in a new app (Android) I am writing. I have communicated with this device without issue with older apps that used https://www.npmjs.com/package/@ionic-native/ble. But when I use this package, when I try to write(), I get an error: "Writing characteristic failed with status code 200.". I have seen other posts that suggest that the combination service / characteristic I am using is not writeable. However, it absolutely is writable as I have done so with other programs and other packages. To be clear I first register for notifications with the service and characteristic below, and then I start writing to it as well. I am able to register for the notifications without any issues. Please let me know if you have any suggestions or ideas as to why this is not working. I can also provide additional information if necessary. Thanks!

More information:
service: "BC2F4CC6-AAEF-4351-9034-D66268E328F0"
characteristic: "06D1E5E7-79AD-4A71-8FAA-373789F7D93C"
Here is the relevant output from getServices():
{
    "uuid": "bc2f4cc6-aaef-4351-9034-d66268e328f0",
    "characteristics": [
        {
            "uuid": "06d1e5e7-79ad-4a71-8faa-373789f7d93c",
            "properties": {
                "broadcast": false,
                "read": false,
                "writeWithoutResponse": false,
                "write": false,
                "notify": true,
                "indicate": false,
                "authenticatedSignedWrites": false,
                "extendedProperties": false
            },
            "descriptors": [
                {
                    "uuid": "00002902-0000-1000-8000-00805f9b34fb"
                },
                {
                    "uuid": "00002901-0000-1000-8000-00805f9b34fb"
                }
            ]
        },
        {
            "uuid": "06d1e5e7-79ad-4a71-8faa-373789f7d93c",
            "properties": {
                "broadcast": false,
                "read": false,
                "writeWithoutResponse": true,
                "write": true,
                "notify": false,
                "indicate": false,
                "authenticatedSignedWrites": false,
                "extendedProperties": false
            },
            "descriptors": [
                {
                    "uuid": "00002901-0000-1000-8000-00805f9b34fb"
                }
            ]
        },
        {
            "uuid": "06d1e5e7-79ad-4a71-8faa-373789f7d93c",
            "properties": {
                "broadcast": false,
                "read": false,
                "writeWithoutResponse": false,
                "write": false,
                "notify": true,
                "indicate": false,
                "authenticatedSignedWrites": false,
                "extendedProperties": false
            },
            "descriptors": [
                {
                    "uuid": "00002902-0000-1000-8000-00805f9b34fb"
                }
            ]
        }
    ]
}

**Comment by peitschie** [comment]
@BAyotte is this on Android or iOS?

**Reply by BAyotte** [reply]
This is on Android.

**Reply by peitschie** [reply]
The 200 error code on Android maps to [ERROR_GATT_WRITE_NOT_ALLOWED](https://developer.android.com/reference/kotlin/android/bluetooth/BluetoothStatusCodes).

This particular flow is new in Android 14+, so potentially something about how the peripheral defines the services and characteristics has a compatibility issue on more recent Android versions.

It's unlike this is a problem in the plugin, unfortunately, as what the plugin does here is very simple.

**Reply by BAyotte** [reply]
I think that error code was added in Android 13 (API 33), although our problem still may be a 14+ related issue. The Android tablet I am using is running Android 14. I have been able to write to that characteristic with our older app that uses the Cordova BLE plugin on this same tablet. The older app that uses this bluetooth plugin https://www.npmjs.com/package/@ionic-native/ble / https://github.com/don/cordova-plugin-ble-central is also building with API 34 (Android 14) and it does not error out when writing to that characteristic. 

We have noticed the service and characteristic (in the original comment) does appear strangely as it has one service with three characteristics that all have the same UUID. I have seen other comments on the internet that mentioned something about how characteristics can have the same UUID but that there is some index under the hood that is different (https://github.com/PhilipsHue/flutter_reactive_ble/issues/478). They asked for a service/characteristic index to be exposed to help solve their problem which seems very similar to ours. Is that possible here as well?

Are there any other tests that would be useful to try?

**Reply by peitschie** [reply]
Ah! Multiple characteristics with the same UUID would definitely make this interesting.

There's an existing request for this in https://github.com/capacitor-community/bluetooth-le/issues/187

I don't have enough technical insight to guess at the difficulty right now though. I'm sure patches attempting this would be welcome however.

**Comment by stanislavnastasiu-dev** [comment]
I ran into the same problem and finally, after a lot of trial and error, I ended up filtering the features based on what I needed to do. In my case, I needed to write, so I filtered the ones that had write=true


```javascript
let bleService: BleService[] = await BleClient.getServices(deviceId);

    const writableServices = bleService.filter((service) =>
      service.characteristics.some(
        (characteristic) => characteristic.properties.write === true
      )
    );

    if (writableServices.length > 0) {
       //save uuid and use on write call 
    }
 ```
